content/machine-learning/regreg.Rmarkdown

```diff
@@ -0,0 +1,129 @@
+
+```{r}
+trainingmat <- as(dfmat_training, "dgCMatrix")
+trainingyvec <- ifelse(dfmat_training$sentiment == "neg", 1, 0)
+
+lasso <- cv.glmnet(x = trainingmat,
+                   y = trainingyvec,
+                   alpha = 1,
+                   nfold = 5,
+                   family = "binomial")
+```
+
+As an initial evaluation of the model, let's take a look at the most predictive features. We begin by obtaining the best value of the penalty parameter $\lambda$:
+
+```{r}
+best.lambda <- which(lasso$lambda == lasso$lambda.min)
+beta <- lasso$glmnet.fit$beta[, best.lambda]
+```
+
+We can now look at the most predictive features for this value of $\lambda$:
+
+```{r}
+mostpreddf <- data.frame(
+    coef = as.numeric(beta),
+    word = names(beta)) %>% 
+  arrange(-coef)
+
+head(mostpreddf,10)
+```
+
+`predict.glmnet` can only take features into consideration that occur both in the training set and the test set, but we can make the features identical by passing `training_dfm` to `dfm_match()` as a pattern.
+
+```{r}
+dfmat_matched <- dfm_match(dfmat_test, features = featnames(dfmat_training))
+```
+
+Next we obtain predicted probabilities for each review in the test set, again converting the document-feature matrix to a sparse matrix:
+
+```{r}
+testmat <- as(dfmat_matched, "dgCMatrix")
+
+preds <- predict(lasso, testmat, type = "response", s = lasso$lambda.min)
+```
+
+Let's inspect how well the classification worked.
+
+```{r}
+actual_class <- ifelse(dfmat_matched$sentiment == "neg", 1, 0) 
+predicted_class <- predict(lasso, testmat, type = "class") %>% as.numeric()
+tab_class <- table(actual_class, predicted_class)
+tab_class
+```
+
+From the cross-table we see that the model slightly underpredicts negative reviews, i.e. produces slightly more false negatives than false positives, but most reviews are correctly predicted. 
+
+We can use the function `confusionMatrix()` from the **caret** package to quantify the performance of the classification.
+
+```{r}
+confusionMatrix(tab_class, mode = "everything")
+```
+
+{{% notice note %}}
+Precision, recall and the F1 score are frequently used to assess the classification performance. Precision is measured as `TP / (TP + FP)`, where `TP` are the number of true positives and  `FP`  the false positives. Recall divides the false positives by the sum of true positives and false negatives `TP / (TP + FN)`. Finally, the F1 score is a harmonic mean of precision and recall `2 * (Precision * Recall) / (Precision + Recall)`.
+{{% /notice %}}
+
+{{% notice ref %}}
+- Jurafsky, Daniel, and James H. Martin. 2018. [_Speech and Language Processing. An Introduction to Natural Language Processing, Computational Linguistics, and Speech Recognition_](https://web.stanford.edu/~jurafsky/slp3/4.pdf). Draft of 3rd edition, September 23, 2018 (Chapter 4). 
```

Comment:
Please list two items in bullet points.

Reply:
This is identical to the Naïve Bayes tutorial, which I based this on. Maybe I should just omit this reference then?

Reply:
I will updated Naive Bayes then. Please keep references.